File: modelAPI/Baidu/BaiduAPI.py
import requests
import json
from collections import deque
from .Token.TokenAllocate import GetToken


class BaiduAPI:
    def __init__(self) -> None:
        token = GetToken()
        self.API_KEY = token['API_KEY']
        self.SECRET_KEY = token['SECRET_KEY']
        self.APP_ID = token['APP_ID']
        self.access_token = self.get_access_token()
        self.headers = {
            'Content-Type': 'application/json'
        }
        self.url = ''

        self.maxLenMessage = 50
        self.messageQueue = deque(maxlen=self.maxLenMessage)
        
    def add_message(self, mesg:str, role="user"):
        self.messageQueue.append({'role': role, 'content': mesg})
        return {"messages": list(self.messageQueue)}
    
    def invoke(self, mesg:str, role="user"):
        self.add_message(mesg,role)
        if len(self.messageQueue) == 0:
            return None
        if self.url == '':
            raise Exception('No url')
        
        message = {"messages": list(self.messageQueue)}
        print(message)
        payload = json.dumps(message)
        response = requests.request("POST", self.url, headers=self.headers, data=payload)
        return response.json()
    
    def response(self, mesg:str, role="user"):
        resp_json = self.invoke(mesg, role)
        print(resp_json)
        return resp_json['result']

    def get_access_token(self):
        """
        使用 AK，SK 生成鉴权签名（Access Token）
        :return: access_token，或是None(如果错误)
        """
        url = "https://aip.baidubce.com/oauth/2.0/token"
        params = {
            "grant_type": "client_credentials",
            "client_id": self.API_KEY,
            "client_secret": self.SECRET_KEY
        }
        return str(requests.post(url, params=params).json().get("access_token"))

File: modelAPI/Baidu/BaiduAPI.go
package Baidu

import (
	"encoding/json"
	"errors"
	"fmt"
	"io/ioutil"
	"llmguard/modelAPI/Baidu/Token"
	"net/http"
	"strings"
)

type Baidu struct {
	ApiKey        string
	SecretKey     string
	AppId         string
	temperature   float64
	accessToken   string
	url           string
	maxLenMessage int
	messageQueue  []map[string]string
	headers       http.Header
}

func NewBaidu() *Baidu {
	token := Token.GetTokens().GetKeys()
	return &Baidu{
		ApiKey:        token.ApiKey,
		SecretKey:     token.SecretKey,
		AppId:         token.AppId,
		accessToken:   "",
		url:           "",
		headers:       http.Header{"Content-Type": {"application/json"}},
		maxLenMessage: 50,
		messageQueue:  make([]map[string]string, 0, 50),
	}
}

func (b *Baidu) AddMessage(mesg string, role string) string {
	if len(b.messageQueue) >= b.maxLenMessage {
		b.messageQueue = b.messageQueue[1:]
	}
	m := map[string]string{"role": role, "content": mesg}
	b.messageQueue = append(b.messageQueue, m)
	message := map[string]interface{}{
		"messages": b.messageQueue,
	}
	jsonMessage, err := json.Marshal(message)
	if err != nil {
		fmt.Println("Error:", err)
	}
	return string(jsonMessage)
}

func (b *Baidu) Invoke(mesg string, role string) (json.RawMessage, error) {
	message := b.AddMessage(mesg, role)
	if len(b.messageQueue) == 0 {
		return nil, errors.New("No message")
	}
	if b.url == "" {
		return nil, errors.New("no url specified")
	}
	req, err := http.NewRequest("POST", b.url, strings.NewReader(message))
	if err != nil {
		return nil, err
	}
	req.Header = b.headers
	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	bodyBytes, err := ioutil.ReadAll(resp.Body)
	if err != nil || resp.StatusCode != 200 {
		return nil, err
	}
	return bodyBytes, nil
}

func (b *Baidu) Response(mesg string, role string) (string, error) {
	jsonStr, err := b.Invoke(mesg, role)
	if err != nil {
		return "", err
	}
	var resp map[string]interface{}
	err = json.Unmarshal([]byte(jsonStr), &resp)
	if err != nil {
		return "", errors.New("result not found or is not a string")
	}

	result := resp["result"].(string)
	return string(result), nil
}

/**
 * 使用 AK，SK 生成鉴权签名（Access Token）
 * @return string 鉴权签名信息（Access Token）
 */
func (b *Baidu) getAccessToken() (string, error) {
	url := "https://aip.baidubce.com/oauth/2.0/token"
	postData := fmt.Sprintf("grant_type=client_credentials&client_id=%s&client_secret=%s", b.ApiKey, b.SecretKey)
	resp, err := http.Post(url, "application/x-www-form-urlencoded", strings.NewReader(postData))
	if err != nil {
		fmt.Println(err)
		return "", err
	}
	defer resp.Body.Close()
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		fmt.Println(err)
		return "", err
	}
	accessTokenObj := map[string]string{}
	json.Unmarshal([]byte(body), &accessTokenObj)
	return accessTokenObj["access_token"], nil
}
